File: Problems/Algorithms/953. Verifying an Alien Dictionary/AlienDictionary.kt
class Solution {
    fun isAlienSorted(words: Array<String>, order: String): Boolean {
        val n = words.size
        if (n == 1) return true
        
        val orderMap = IntArray(26) { 0 }
        for (i in 0..25) {
            orderMap[order.get(i)-'a'] = i
        }
        
        for (i in 0..n-2) {
            val curr = words[i]
            val next = words[i+1]
            for (j in 0..words[i].length-1) {
                if (j >= next.length) return false
                
                if (orderMap[curr.get(j)-'a'] > orderMap[next.get(j)-'a']) {
                    return false
                } else if (orderMap[curr.get(j)-'a'] < orderMap[next.get(j)-'a']) {
                    break
                }
            }
        }
        
        return true
    }
}

File: Problems/Algorithms/953. Verifying an Alien Dictionary/alien_dictionary.rs
impl Solution {
    pub fn is_alien_sorted(words: Vec<String>, order: String) -> bool {
        let order_map = order.as_bytes()
            .iter()
            .enumerate()
            .fold([0; 26], |mut acc, (i, c)| {
                acc[(c-b'a') as usize] = i;
                acc
            });
        
        words.iter()
            .map(|s| {
                s.as_bytes()
                    .iter()
                    .map(|&c| order_map[(c-b'a') as usize])
                    .collect::<Vec<_>>()
            })
            .collect::<Vec<_>>()
            .windows(2)
            .all(|pairs| pairs[0] <= pairs[1])
    }
}

